src/nvim/message.c

```diff
@@ -3145,6 +3167,71 @@ void msg_check(void)
   }
 }
 
+bool msg_enable_ext(void)
+{
+  return ui_has(kUIMessages) || msg_enable_msgfunc();
+}
+
+bool msg_enable_msgfunc(void)
+{
+  // msgfunc is disabled in command line mode
+  // Because default echo is used for command line redraw
+  return *p_msgfunc != NUL && !msg_check_loop
+    && !redir_off && !need_wait_return
+    && State != CMDLINE && State != CONFIRM;
```

Comment:
FWIW, according to the docs of vim.h

https://github.com/neovim/neovim/blob/828bf128a64466f254629b102e283af35666cd05/src/nvim/vim.h#L43-L44

I suggested `State & CMDLINE` because that is how it's used in other places, e.g.

https://github.com/neovim/neovim/blob/828bf128a64466f254629b102e283af35666cd05/src/nvim/aucmd.c#L106

https://github.com/neovim/neovim/blob/725cbe7d414f609e769081276f2a034e32a4337b/src/nvim/cursor_shape.c#L332

I don't know if `!= CMDLINE` is as effective.